### ISSUE TITLE
[TFLite] TF MatMul operator converted to an unsupported TFL 16x16 FullyConnected operator with 16x8 post-training quantization

### ISSUE THREAD

**Issue by Tessil** [tensorflow/tensorflow#44707]
Hello,

When using the `tf.linalg.matmul` function with TFLite 16-bit post-training quantization the operation is sometimes transformed into a `TFL::FullyConnectedOp` operator with 16-bit inputs and 16-bit weights. Unfortunately the 16-bit version of the FC operator only supports 16-bit inputs and 8-bit weights.

The conversion occurs in two places depending on how `tf.linalg.matmul` is interpreted (as a `TF::MatMulOp` operator or a `TF::BatchMatMul(V2)Op`):
* The `TF::MatMulOp` is converted to a `TFL::FullyConnectedOp` in [legalize_tf.cc](https://github.com/tensorflow/tensorflow/blob/c86b90dc9e3dc51dfc95964449570160cb12c1fe/tensorflow/compiler/mlir/lite/transforms/legalize_tf.cc#L242).
* The `TF::BatchMatMul(V2)Op` on the other hand is converted in [unroll_batch_matmul.cc](https://github.com/tensorflow/tensorflow/blob/c86b90dc9e3dc51dfc95964449570160cb12c1fe/tensorflow/compiler/mlir/tensorflow/transforms/unroll_batch_matmul.cc#L190) to a `TF::MatMulOp` which is then converted by the previous transformation into a `TFL::FullyConnectedOp`. This conversion pass is only enabled when [`PrepareTFPass::unfold_batch_matmul_`](https://github.com/tensorflow/tensorflow/blob/c86b90dc9e3dc51dfc95964449570160cb12c1fe/tensorflow/compiler/mlir/lite/transforms/prepare_tf.cc#L1214) is true.

To disable the second conversion we could add a check on the `toco_flags.inference_type()` flag in [graphdef_to_tfl_flatbuffer.cc](https://github.com/tensorflow/tensorflow/blob/c86b90dc9e3dc51dfc95964449570160cb12c1fe/tensorflow/compiler/mlir/lite/python/graphdef_to_tfl_flatbuffer.cc#L88) (and other appropriate places) and set `pass_config.unfold_batch_matmul` to false when the inference type is int16. The `TF::BatchMatMul(V2)Op` would then be converted to a `TFL::BatchMatMulOp` which support 16x16 inputs without trouble.

The problem remains though if we have a simple 16x16 `TF::MatMulOp` operator. We can't just disable the conversion in the same way as for `TF::BatchMatMul(V2)Op` as there isn't any `TFL::MatMulOp`. 

One option would be to disable this transformation in int16 post-training quantization in the same way as for `TF::BatchMatMul(V2)Op` and add a `TF::MatMulOp` to `TF::BatchMatMul(V2)Op` conversion pass. The `TFL::BatchMatMulOp`  would then be used for the cases where we didn't converted the `TF::MatMulOp` to a `TFL::FullyConnectedOp`.

Thibaut

**Comment by jdduke** [comment]
@talumbau @karimnosseir thoughts on addressing this?

**Comment by karimnosseir** [comment]
I am confused with the details you mention. If you're using post training quantization then the ops should be a floating point not int16. The int16 quantization happens after the conversion which you point too.

On the other hand
Since TFLite kernel doesn't support filter with int16 type, then we should remove the type from here
https://github.com/tensorflow/tensorflow/blob/c86b90dc9e3dc51dfc95964449570160cb12c1fe/tensorflow/compiler/mlir/lite/ir/tfl_ops.td#L950
Then if your model has Matmul with int16 type originally it will not be converted to TFL, and will remain as TF op since the kernel doesn't support it.

**Comment by Tessil** [comment]
I modified the title as it was a bit confusing. The MatMul op is initially in floating-point when the conversion to `TFL::FullyConnectedOp` happens. It's later during the post-training quantization that the `TFL::FullyConnectedOp` is transformed to an unsupported 16x16 version.

The https://github.com/tensorflow/tensorflow/pull/44735 PR solved the second problem and disabled the unfolding of the `TF::BatchMatMul(V2)Op` when 16x8 post-training quantization is used.

The problem remains if we initially have a floating-point `TF::MatMulOp` which is [converted](https://github.com/tensorflow/tensorflow/blob/c86b90dc9e3dc51dfc95964449570160cb12c1fe/tensorflow/compiler/mlir/lite/transforms/legalize_tf.cc#L242) to a floating-point `TFL::FullyConnectedOp`  which is then converted to an unsupported 16x16 `TFL::FullyConnectedOp` during the quantization.

If we modify a bit the example of #44735 PR to have a `TF::MatMulOp` instead of a `TF::BatchMatMulV2Op` (a bit contrived example but it's to illustrate the problem):
```python
import numpy as np
import tensorflow as tf

input_shape = [4, 2]
input1 = tf.keras.Input(shape=input_shape, batch_size=1)
input2 = tf.keras.Input(shape=input_shape, batch_size=1)
output = tf.keras.layers.Lambda(lambda x: tf.linalg.matmul(
                                   x[0][0], x[1][0], transpose_b=True)
                                )([input1, input2])
model = tf.keras.Model(inputs=[input1, input2], outputs=output)


def get_rand_date():
    return np.random.rand(1, *input_shape).astype(np.float32)

def representative_data_gen():
    yield [get_rand_date(), get_rand_date()]

converter = tf.lite.TFLiteConverter.from_keras_model(model)
converter.representative_dataset = representative_data_gen
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.target_spec.supported_ops = [
    tf.lite.OpsSet.EXPERIMENTAL_TFLITE_BUILTINS_ACTIVATIONS_INT16_WEIGHTS_INT8]

tflite_model = converter.convert()
interpreter = tf.lite.Interpreter(model_content=tflite_model)
interpreter.allocate_tensors()
interpreter.set_tensor(interpreter.get_input_details()[0]["index"],
                       get_rand_date())
interpreter.set_tensor(interpreter.get_input_details()[1]["index"],
                       get_rand_date())
interpreter.invoke()
```

We end-up with the following error: `RuntimeError: tensorflow/lite/kernels/fully_connected.cc:123 input->type != kTfLiteFloat32 (INT16 != FLOAT32)Node number 4 (FULLY_CONNECTED) failed to prepare.`

This problem is more minor for us (the `TF::BatchMatMulV2Op`  was more problematic) but it may be worth looking into it.

**Comment by karimnosseir** [comment]
Ethan, Can you please triage and check if PTQ still introduce this

Thanks

**Comment by Tessil** [comment]
Solved by 0546868a385cb10028f831437980934299e3d13b as we can now use `_experimental_disable_batchmatmul_unfold=True`. Thanks.
